### PR TITLE
Public Api with Oauth Protection

### PR DESCRIPTION
**Pull request by rohitsharma04**
This PR doesn't have policy handling endpoints as those endpoints require policy validation code from #449
I'll do a different PR for that after these #449 and this PR both are merged.

### CODE REVIEW

File: src/autoscaler/api/publicapiserver/oauth.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+}
+
+func (oam *OAuthMiddleware) getUserScope(userToken string, userId string) ([]string, error) {
+	tokenEndpoint, err := oam.getCFTokenEndpoint()
+	if err != nil {
+		return nil, err
+	}
+
+	parameters := url.Values{}
+	parameters.Add("token", strings.Split(userToken, " ")[1])
```
Comment (boyang9527) [inline src/autoscaler/api/publicapiserver/oauth.go]: check the token to avoid out of range

File: src/autoscaler/api/publicapiserver/oauth.go [hunk 2]
```diff
@@ -0,0 +1,242 @@
+package publicapiserver
```
Comment (boyang9527) [inline src/autoscaler/api/publicapiserver/oauth.go]: we can move all this to cf package.

File: src/autoscaler/api/publicapiserver/public_api_handler.go [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+		h.logger.Error("Failed to info file", err, lager.Data{"info-file-path": h.conf.InfoFilePath})
+		handlers.WriteJSONResponse(w, http.StatusInternalServerError, models.ErrorResponse{
+			Code:    "Interal-Server-Error",
+			Message: "Failed to load info"})
+		return
+	}
+	w.Write([]byte(info))
+}
+
+func (h *PublicApiHandler) GetHealth(w http.ResponseWriter, r *http.Request, vars map[string]string) {
```
Comment (boyang9527) [inline src/autoscaler/api/publicapiserver/public_api_handler.go]: looks like we always return "alive" then why we need it?

Comment (boyang9527) [reply src/autoscaler/api/publicapiserver/public_api_handler.go]: I would assume this should be prometheus.

**Comment by cdlliuy** [comment]
@rohitsharma04 ,  will you add CURD of policy to public API endpoint in another PR?

File: src/autoscaler/api/publicapiserver/public_api_handler.go [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+			Message: "metrictype is required",
+		})
+		return
+	}
+
+	path, _ := routes.EventGeneratorRoutes().Get(routes.GetAggregatedMetricHistoriesRouteName).URLPath("appid", appId, "metrictype", metricType)
+
+	url := h.conf.EventGenerator.EventGeneratorUrl + path.RequestURI() + "?" + parameters.Encode()
+
+	resp, err := h.metricsCollectorClient.Get(url)
```
Comment (cdlliuy) [inline src/autoscaler/api/publicapiserver/public_api_handler.go]: use eventgenClient?

File: src/autoscaler/api/publicapiserver/public_api_handler.go [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+	handlers.WriteJSONResponse(w, resp.StatusCode, paginatedResponse)
+}
+
+func (h *PublicApiHandler) GetInstanceMetricsHistories(w http.ResponseWriter, r *http.Request, vars map[string]string) {
+	appId := vars["appId"]
+
+	metricType := vars["metricType"]
+	instanceIndex := r.URL.Query().Get("instance-index")
+
+	h.logger.Info("Get ScalingHistories", lager.Data{"appId": appId, "metricType": metricType, "instanceIndex": instanceIndex})
```
Comment (cdlliuy) [inline src/autoscaler/api/publicapiserver/public_api_handler.go]: Get InstanceMetricsHistories

File: src/autoscaler/api/publicapiserver/public_api_handler.go (hunk 3 above)
Comment (cdlliuy) [reply src/autoscaler/api/publicapiserver/public_api_handler.go]: it is a legacy design issue.  Before moving to Prometheus,   there is a /health endpoint on api server to show uptime.  Also  autoscaler CLI is using the /health endpoint to determine whether the api server is accessible.  i.e.  CLI expects to get 200 OK with /health
https://github.com/cloudfoundry-incubator/app-autoscaler-cli-plugin/blob/master/src/cli/api/apihelper.go#L140

File: src/autoscaler/api/publicapiserver/oauth.go (hunk 2 above)
Comment (rohitsharma04) [reply src/autoscaler/api/publicapiserver/oauth.go]: done

**Comment by cdlliuy** [comment]
@rohitsharma04 ,  could you please resolve the conflict?

**Comment by rohitsharma04** [comment]
> @rohitsharma04 , could you please resolve the conflict?

@cdlliuy  done.

File: src/autoscaler/api/cmd/api/main.go [hunk 6]
```diff
@@ -48,31 +51,49 @@ func main() {
...
-		logger.Error("failed to connect bindingdb database", err, lager.Data{"dbConfig": conf.DB.BindingDB})
+		logger.Error("failed to connect to policydb database", err, lager.Data{"dbConfig": conf.DB.PolicyDB})
 		os.Exit(1)
 	}
-	defer bindingDB.Close()
+	defer policyDb.Close()
+
+	if !conf.UseBuildInMode {
+		var bindingDB db.BindingDB
+		bindingDB, err = sqldb.NewBidingSQLDB(conf.DB.BindingDB, logger.Session("bindingdb-db"))
```
Comment (cdlliuy) [inline src/autoscaler/api/cmd/api/main.go]: BidingDB?  typo?

**Comment by cdlliuy** [comment]
@rohitsharma04 ,  I guess you will have another PR to add policy CRUD in public API server, right?   If so,  this PR is LGTM

**Comment by rohitsharma04** [comment]
> @rohitsharma04 , I guess you will have another PR to add policy CRUD in public API server, right? If so, this PR is LGTM

yes.

**Comment by paltanmoy** [comment]
@rohitsharma04  Please create the separate PR for policy CRUD. 

LGTM

**Pull request merged**